internal/gatewayapi/securitypolicy.go

```diff
@@ -757,25 +907,129 @@ func (t *Translator) translateSecurityPolicyForGateway(
+	}
+	if gateway.Gateway == nil {
+		return fmt.Errorf("translateSecurityPolicyForGateway: embedded Gateway is nil for gateway context %v (policy %s/%s)", gateway, policy.Namespace, policy.Name)
+	}
+
 	irKey := t.getIRKey(gateway.Gateway)
 	// Should exist since we've validated this
-	x := xdsIR[irKey]
+	x, ok := xdsIR[irKey]
+	if !ok || x == nil {
+		return fmt.Errorf("no IR found for gateway %s (cannot apply SecurityPolicy)", irKey)
+	}
+	sectionName := ""
+	if target.SectionName != nil {
+		sectionName = string(*target.SectionName)
+	}
+	// Detect protocol composition of the Gateway (needed for mixed whole-gateway handling)
+	hasHTTP, hasTCP := false, false
+	for _, l := range gateway.Spec.Listeners {
+		switch l.Protocol {
+		case gwapiv1.HTTPProtocolType, gwapiv1.HTTPSProtocolType:
+			hasHTTP = true
+		case gwapiv1.TCPProtocolType:
+			hasTCP = true
+		}
+	}
+	isMixedWholeGateway := sectionName == "" && hasHTTP && hasTCP
+	// Detect if target is TCP listener (or all listeners TCP when sectionName empty).
+	isTCPListener := false
+	if sectionName != "" {
+		// defensive loop: ensure gateway.Spec is available
+		for _, l := range gateway.Spec.Listeners {
+			if string(l.Name) == sectionName && l.Protocol == gwapiv1.TCPProtocolType {
+				isTCPListener = true
+				break
+			}
+		}
+	} else {
+		allTCP := true
+		for _, l := range gateway.Spec.Listeners {
+			if l.Protocol != gwapiv1.TCPProtocolType {
+				allTCP = false
+				break
+			}
+		}
+		isTCPListener = allTCP
+	}
 
 	policyTarget := irStringKey(policy.Namespace, string(target.Name))
+
+	if isTCPListener {
+		// Apply ONLY Authorization to all TCP routes under the targeted listener(s).
+		for _, tl := range x.TCP {
+			if tl == nil {
+				continue
+			}
+			// defensive: ensure metadata present
+			// Listener name format: namespace/gatewayName/listenerName
+			// compute effective section name from metadata if present, else fallback to name suffix
+			effectiveSection := ""
+			if tl.Metadata != nil && tl.Metadata.SectionName != "" {
+				effectiveSection = tl.Metadata.SectionName
+			} else {
+				// fallback: last path segment of tl.Name (`namespace/gateway/listener`)
+				// TODO: remove fallback once IR builder guarantees tl.Metadata.SectionName is populated.
+				if idx := strings.LastIndex(tl.Name, "/"); idx >= 0 && idx < len(tl.Name)-1 {
+					effectiveSection = tl.Name[idx+1:]
+				} else {
+					effectiveSection = tl.Name
+				}
+			}
+
+			if t.MergeGateways && !strings.HasPrefix(tl.Name, policyTarget) {
+				continue
+			}
+			if sectionName != "" && effectiveSection != sectionName {
+				continue
+			}
+			for _, r := range tl.Routes {
+				if r == nil {
+					continue
+				}
+				if r.Security != nil {
+					continue
+				}
+				r.Security = &ir.SecurityFeatures{
+					Authorization: authorization,
+				}
+			}
+		}
+		return errs
+	}
+
+	// HTTP branch: defensive nil checks for listeners/metadata
```

Comment:
can we avoid these changes in this PR ? are they needed to enable TCPRoute SP support ? they are adding more complexity to this diff